src/plugins/data/public/ui/typeahead/suggestions_component.tsx

```diff
@@ -141,14 +143,130 @@ export default class SuggestionsComponent extends Component<SuggestionsComponent
+  const isSuggestionsListFittable =
+    documentHeight - (containerRect.top + containerRect.height) >
+    SUGGESTIONS_LIST_REQUIRED_BOTTOM_SPACE;
+  const verticalListPosition = isSuggestionsListFittable
+    ? `top: ${pageYOffset + containerRect.bottom - SUGGESTIONS_LIST_REQUIRED_TOP_OFFSET}px;`
+    : `bottom: ${documentHeight - (pageYOffset + containerRect.top)}px;`;
+
+  return (
+    <StyledSuggestionsListDiv
+      left={containerRect.left}
+      width={containerRect.width}
+      verticalListPosition={verticalListPosition}
+    >
+      <div
+        className={classNames('kbnTypeahead', {
+          'kbnTypeahead--small': props.suggestionsSize === 's',
+        })}
+      >
+        <div
+          className={classNames('kbnTypeahead__popover', {
+            ['kbnTypeahead__popover--bottom']: isSuggestionsListFittable,
+            ['kbnTypeahead__popover--top']: !isSuggestionsListFittable,
+          })}
+        >
+          {children(containerRect)}
+        </div>
+      </div>
+    </StyledSuggestionsListDiv>
+  );
+});
+
+function useDimensions(
+  container: HTMLElement | null
+): [{ documentHeight: number }, { pageYOffset: number; pageXOffset: number }, DOMRect | null] {
+  const [documentHeight, setDocumentHeight] = useRafState(
+    () => document.documentElement.clientHeight || window.innerHeight
+  );
+
+  const [pageOffset, setPageOffset] = useRafState<{ pageXOffset: number; pageYOffset: number }>(
+    () => ({
+      pageXOffset: window.pageXOffset,
+      pageYOffset: window.pageYOffset,
+    })
+  );
+
+  const [containerRect, setContainerRect] = useRafState<DOMRect | null>(() => {
+    return container?.getBoundingClientRect() ?? null;
+  });
+
+  const updateContainerRect = React.useCallback(() => {
+    setContainerRect((oldRect: DOMRect | null) => {
+      const newRect = container?.getBoundingClientRect() ?? null;
+      const rectsEqual = shallowEqual(oldRect?.toJSON(), newRect?.toJSON());
+      return rectsEqual ? oldRect : newRect;
+    });
+  }, [container, setContainerRect]);
+
+  React.useEffect(() => {
+    const handler = () => {
+      setDocumentHeight(document.documentElement.clientHeight || window.innerHeight);
+    };
+
+    window.addEventListener('resize', handler, { passive: true });
+
+    return () => {
+      window.removeEventListener('resize', handler);
+    };
+  }, [setDocumentHeight]);
+
+  React.useEffect(() => {
+    const handler = () => {
+      setPageOffset((state) => {
+        const { pageXOffset, pageYOffset } = window;
+        return state.pageXOffset !== pageXOffset || state.pageYOffset !== pageYOffset
+          ? {
+              pageXOffset,
+              pageYOffset,
+            }
+          : state;
+      });
+
+      updateContainerRect();
+    };
+
+    window.addEventListener('scroll', handler, { passive: true, capture: true });
+
+    const resizeObserver =
+      typeof window.ResizeObserver !== 'undefined' &&
```

Comment:
Agree that support is pretty good, but I was actually deciding between: 

A. Using it only when available because this isn't critical functionality (so if it is not available, everything is still working, but with some position bugs) 
B. Using `resize-observer-polyfill` like we do in other places. But I wanted to avoid including couple more Kbs into this async chunk for this - https://bundlephobia.com/package/resize-observer-polyfill@1.5.1
C. Using it directly assuming it always exists. I didn't go with it because it wasn't necessary to introduce this breaking change. As it might break someone who is using Kibana on some TV platform or older Safari without a good need for it. But since we are actually using it in some places directly, I created: https://github.com/elastic/kibana/issues/121509 to address this separately